src/app/collective-rewards/hooks/builderHooks.ts

```diff
@@ -0,0 +1,125 @@
+import { BuilderRegistryAbi } from '@/lib/abis/v2/BuilderRegistryAbi'
+import { SponsorsManagerAddress } from '@/lib/contracts'
+import { Address, zeroAddress } from 'viem'
+import { useReadContract, useWaitForTransactionReceipt, useWriteContract } from 'wagmi'
+import { GaugeAbi } from '@/lib/abis/v2/GaugeAbi'
+import { createZeroAddressError } from '@/shared/errors/zeroAddressError'
+import { useEffect, useMemo, useRef } from 'react'
+import { useAlertContext } from '@/app/providers'
+
+export const useGetGauge = (builder: Address) =>
+  useReadContract({
+    abi: BuilderRegistryAbi,
+    address: SponsorsManagerAddress,
+    functionName: 'builderToGauge',
+    args: [builder as Address],
+  })
+
+export const useClaimAllRewards = (builder: Address) => {
+  const { writeContractAsync, error: executionError, data: hash, isPending } = useWriteContract()
+  const { data: gauge, isPending: isGaugePending, error: gaugeError, isFetched } = useGetGauge(builder)
+
+  const { isLoading, isSuccess, data, error: receiptError } = useWaitForTransactionReceipt({ hash })
+
+  const isClaimFunctionReady = useMemo(
+    () => !isGaugePending && !gaugeError && !!gauge,
+    [isGaugePending, gaugeError, gauge],
+  )
+
+  const error = useMemo(() => {
+    if (gaugeError) {
+      return {
+        ...gaugeError,
+        shortMessage: `Failed getting gauge for builder ${builder} (${gaugeError.message})`,
+      }
+    }
+
+    if (executionError) {
+      return {
+        ...executionError,
+        shortMessage: `Failed claim execution (${executionError.message})`,
+      }
+    }
+
+    if (receiptError) {
+      return {
+        ...receiptError,
+        shortMessage: `Failed to get claim execution receipt (${receiptError.message})`,
+      }
+    }
+
+    if (isFetched && gauge === zeroAddress) {
+      return createZeroAddressError('Gauge', { builder })
```

Comment:
Should we be more specific and say "The gauge doesn't exist"?

Reply:
I thought we could have a unified error for this, as we're going to check for non-zero address in other places, but sure. I can change it to a specific error

Reply:
done